File: onboard/eq2-cameras/src/stream_registry.cpp
#include "stream_registry.hpp"

using namespace equinox::camera;
using namespace peel;

StreamRegistry::StreamRegistry()
{

}
StreamRegistry::~StreamRegistry()
{

}

tl::expected<std::string, equinox::camera::CamErrorDetails> StreamRegistry::create_stream(const std::string uid, const char * name, RefPtr<Gst::Device> device)
{
  auto pipeline = Gst::Pipeline::create(name);
  if (!pipeline) return MAKE_CAM_ERROR_EXPECTED(CamError::PipelineCreationFailed, "Pipeline was unable to be created");
  
  auto float_ptr_src = device->create_element("source");
  if (!float_ptr_src) return MAKE_CAM_ERROR_EXPECTED(CamError::SourceCreationFailed, "Pipeline source was unable to be created");

  peel::RefPtr<peel::Gst::Element> ref_ptr_src = float_ptr_src;  // upcast to ref ptr for pipeline
  pipeline->add(ref_ptr_src); 

  auto stream = std::make_shared<StreamInstance>();

  stream->pipeline = pipeline;
  stream->source = ref_ptr_src;

  auto source = pipeline->get_by_name("source");
  if (source != ref_ptr_src) return MAKE_CAM_ERROR_EXPECTED(CamError::AdditionFailed, "Failed to add source element to pipeline");

  {
    std::lock_guard<std::mutex> lock(stream_mutex_);
    stream_map_.insert_or_assign(uid, std::move(stream));
  }
  return uid;
}

tl::expected<void, equinox::camera::CamErrorDetails> StreamRegistry::remove_stream(const std::string uid)
{
  {
    std::lock_guard<std::mutex> lock(stream_mutex_);
    auto it = stream_map_.find(uid);
  
    if (it == stream_map_.end())
    {
      return MAKE_CAM_ERROR_EXPECTED(CamError::DeviceNotFound, 
        "Could not find unplugged device " + uid + " in registry");
    }
    stream_map_.erase(it);
    spdlog::info("Stream {} successfully removed from registry", uid);

    return {};
  }

}

File: onboard/eq2-cameras/src/stream_registry.rs
use gstreamer as gst;
use gstreamer::prelude::*;

use std::collections::HashMap;
use crate::camera_types::{CamError, StreamInstance};

pub struct StreamRegistry {
    streams: HashMap<String, StreamInstance>,
}

// paused is equivilant to play for streams. either be null, ready or playing.
impl StreamRegistry {
    pub fn new() -> Self {
        Self { streams: HashMap::new() }
    }

    pub fn create_stream(&mut self, uid: &str, device: &gst::Device) -> Result<(), CamError> {
        let pipeline = gst::Pipeline::with_name(uid);
        
        let source = device.create_element(Some("source"))
            .map_err(|_| CamError::ElementCreationFailed("source".into()))?;

        let videoconvert = gst::ElementFactory::make("videoconvert")
            .name(format!("videoconvert_{}", uid))
            .build()
            .map_err(|_| CamError::ElementCreationFailed("videoconvert".into()))?;

        let videoscale = gst::ElementFactory::make("videoscale")
            .name(format!("videoscale_{}", uid))
            .build()
            .map_err(|_| CamError::ElementCreationFailed("videoscale".into()))?;
        
        let autovideosink = gst::ElementFactory::make("autovideosink")
            .name(format!("sink_{}", uid))
            .build()
            .map_err(|_| CamError::ElementCreationFailed("sink".into()))?;
        
        let webrtcbin = gst::ElementFactory::make("webrtcbin")
            .name(format!("webrtc_{}", uid))
            .build()
            .map_err(|_| CamError::ElementCreationFailed("webrtcbin".into()))?;

        pipeline.add_many([source.clone(), videoconvert.clone(), videoscale.clone(), autovideosink.clone()]);

        gst::Element::link_many([source.clone(), videoconvert.clone(), videoscale.clone(), autovideosink.clone()])?;

        pipeline.set_state(gst::State::Playing)
            .map_err(|_| CamError::PipelineError("pipeline".into()))?;

        self.streams.insert(uid.to_string(), StreamInstance {
            pipeline,
            source,
            webrtcbin: gst::ElementFactory::make("webrtcbin").build().unwrap(),
        });

        log::info!("Successfully created stream for UID: {}", uid);
        Ok(())
    }

    pub fn remove_stream(&mut self, uid: &str) {
        if let Some(instance) = self.streams.remove(uid) {
            let _ = instance.pipeline.set_state(gst::State::Null);
            log::info!("Stream {} stopped and removed", uid);
        }
    }

    pub fn setup_stream(&mut self, uid: &str) -> Result<(), CamError> {
        Ok(())
    }

    /// Starts stream based off uid
    pub fn start_stream(&mut self, uid: &str) -> Result<(), CamError> {
        let instance = self.streams.get(uid)
            .ok_or_else(|| CamError::DeviceNotFound(format!("Stream not found for UID: {}", uid)))?; 
            
        instance.pipeline.set_state(gst::State::Playing);
        Ok(())
    }
}